src/Chainweb/SPV/CreateProof.hs

```diff
@@ -248,7 +249,7 @@ outputProofPrefix i db payload = do
     -- 1. TX proof
     Just outs <- tableLookup blockOutputTable $ _blockPayloadOutputsHash payload
         -- TODO: use the transaction tree cache
-    let (!subj, pos, t) = bodyTree @_ @ChainwebHashTag outs i
+    let !(!subj, pos, t) = bodyTree @_ @ChainwebHashTag outs i
```

Comment:
Why does this outer `!` have any meaning? Since you are already pattern matching the scrutinee against a tuple pattern, then it's already going to be in WHNF.

Reply:
Here's an example session, in the REPL, which may explain things:
```
> let (!x, y) = undefined
-- nothing happens, hence the tuple has not been evaluated at all.
-- this is a *lazy binding*
> y
*** Exception: Prelude.undefined
-- ^ now the tuple has been evaluated

> let !(!x, y) = undefined
*** Exception: Prelude.undefined
-- ^ this is a strict binding, having a bang on the outermost level is special. 

From an older version of the GHC user guide: 

"a bang at the top level of a let or where binding makes the binding strict, regardless of the 
pattern... the bang at the top of a binding is not part of the pattern; rather it is part of the 
syntax of the binding, creating a 'bang-pattern binding''

> let (!x, y) = (undefined, 1)
-- nothing happens, `x` has not been evaluated because the entire binding is lazy
> y
*** Exception: Prelude.undefined
-- now that the tuple has been matched against with the pattern (!x, y),
-- `x` must be evaluated because of the bang pattern
```

So my reading of this code is: strictly bind `subj`, `pos` and `t` to the components of the tuple returned by `bodyTree`, and evaluate `subj` to WHNF.

The prior version of this code doesn't trigger evaluation of `subj`, `pos`, or `t`, because:
- `subj` is returned in a lazy tuple. Whereas evaluating `pos` or `t` would trigger evaluating `subj`,
- the only use of `pos` and `t` is stuffing it into a lazy tuple in the head of a lazy non-empty list, `proof`. `proof` is already in WHNF, and the bang on `proof` therefore does nothing.

So we need even more bangs, on `pos` and `t`, and for the bang on `proof` to be deleted.

Reply:
I'll do this in a separate PR instead and leave this be.

Reply:
Sounds good. And I forgot about irrefutable matching.